### ISSUE TITLE
[JENKINS-39968] git checkouts are not pipeline-parallel safe

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#2996]
<p>We have a specific workflow that enables us to discover that bug. However, it could and will probably happen to other people soon.</p>

<p>When running scm checkouts in parallels in a Jenkins file, the changes discovered are discovered for <b>every</b> git checkout that runs at the same time.</p>

<p>In our example, we run checkouts with 8 threads in parallel (background: <a href="https://roidelapluie.be/blog/2016/11/18/gitslave-jenkins/" class="external-link" target="_blank" rel="nofollow noopener">https://roidelapluie.be/blog/2016/11/18/gitslave-jenkins/</a> ).</p>

<p>One change is displayed in 8 different repositories, the 8 repositories that are checked out at the same time.</p>

<p>The next checkouts do not contains the changes, so I guess it is kind of a global variable that is reset after the checkout.</p>

<pre>
lock(<span class="code-quote">'checkout'</span>) {
    def checkouts = [:]
    def threads = 8

    stage(<span class="code-quote">'Super Repo'</span>) {
checkout([$class: <span class="code-quote">'GitSCM'</span>,
branches: [[name: branch]],
doGenerateSubmoduleConfigurations: <span class="code-keyword">false</span>,
extensions: [[$class: <span class="code-quote">'CleanCheckout'</span>], [$class: <span class="code-quote">'ScmName'</span>, name: <span class="code-quote">'<span class="code-keyword">super</span>'</span>]],
submoduleCfg: [],
userRemoteConfigs: [[url: <span class="code-quote">"${gitBaseUrl}/${superRepo}"</span>]]])

def repos = readFile(<span class="code-quote">'.gitslave'</span>)
def reposLines = repos.readLines()
<span class="code-keyword">for</span> (i = 0; i < threads; i++){
    checkouts[<span class="code-quote">"<span class="code-object">Thread</span> ${i}"</span>] = []
}
def idx = 0
<span class="code-keyword">for</span> (line in reposLines) {
    def repoInfo = line.split(<span class="code-quote">' '</span>)
    def repoUrl = repoInfo[0]
    def repoPath = repoInfo[1]
    def curatedRepoUrl = repoUrl.substring(4, repoUrl.length()-1)
    def curatedRepoPath = repoPath.substring(1, repoPath.length()-1)
    echo(<span class="code-quote">"<span class="code-object">Thread</span> ${idx%threads}"</span>)
    checkouts[<span class="code-quote">"<span class="code-object">Thread</span> ${idx%threads}"</span>] << [path: curatedRepoPath, url: <span class="code-quote">"${gitBaseUrl}/${curatedRepoUrl}"</span>]
    idx++
}
    }
    stage(<span class="code-quote">'GitSlave Repos'</span>) {
def doCheckouts = [:]
<span class="code-keyword">for</span> (i = 0; i < threads; i++){
    def j = i
    doCheckouts[<span class="code-quote">"<span class="code-object">Thread</span> ${j}"</span>] = {
<span class="code-keyword">for</span> (co in checkouts[<span class="code-quote">"<span class="code-object">Thread</span> ${j}"</span>]) {
    retry(3) {
checkout([$class: <span class="code-quote">'GitSCM'</span>,
branches: [[name: branch]],
doGenerateSubmoduleConfigurations: <span class="code-keyword">false</span>,
extensions: [[$class: <span class="code-quote">'RelativeTargetDirectory'</span>, relativeTargetDir: co.path], [$class: <span class="code-quote">'CleanCheckout'</span>], [$class: <span class="code-quote">'ScmName'</span>, name: co.path]],
submoduleCfg: [],
userRemoteConfigs: [[url: co.url]]])
    }
}
    }
}
parallel doCheckouts
    }
</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/roidelapluie.png" title="roidelapluie's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=roidelapluie">roidelapluie</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968" target="_blank">git checkouts are not pipeline-parallel safe</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin, workflow-scm-step-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 5
<li><b>watchers</b>: 10
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We have a specific workflow that enables us to discover that bug. However, it could and will probably happen to other people soon.</p>

<p>When running scm checkouts in parallels in a Jenkins file, the changes discovered are discovered for <b>every</b> git checkout that runs at the same time.</p>

<p>In our example, we run checkouts with 8 threads in parallel (background: <a href="https://roidelapluie.be/blog/2016/11/18/gitslave-jenkins/" class="external-link" target="_blank" rel="nofollow noopener">https://roidelapluie.be/blog/2016/11/18/gitslave-jenkins/</a> ).</p>

<p>One change is displayed in 8 different repositories, the 8 repositories that are checked out at the same time.</p>

<p>The next checkouts do not contains the changes, so I guess it is kind of a global variable that is reset after the checkout.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock(<span class="code-quote">'checkout'</span>) {
    def checkouts = [:]
    def threads = 8

    stage(<span class="code-quote">'Super Repo'</span>) {
        checkout([$class: <span class="code-quote">'GitSCM'</span>,
                branches: [[name: branch]],
                doGenerateSubmoduleConfigurations: <span class="code-keyword">false</span>,
                extensions: [[$class: <span class="code-quote">'CleanCheckout'</span>], [$class: <span class="code-quote">'ScmName'</span>, name: <span class="code-quote">'<span class="code-keyword">super</span>'</span>]],
                submoduleCfg: [],
                userRemoteConfigs: [[url: <span class="code-quote">"${gitBaseUrl}/${superRepo}"</span>]]])

        def repos = readFile(<span class="code-quote">'.gitslave'</span>)
        def reposLines = repos.readLines()
        <span class="code-keyword">for</span> (i = 0; i &lt; threads; i++){
            checkouts[<span class="code-quote">"<span class="code-object">Thread</span> ${i}"</span>] = []
        }
        def idx = 0
        <span class="code-keyword">for</span> (line in reposLines) {
            def repoInfo = line.split(<span class="code-quote">' '</span>)
            def repoUrl = repoInfo[0]
            def repoPath = repoInfo[1]
            def curatedRepoUrl = repoUrl.substring(4, repoUrl.length()-1)
            def curatedRepoPath = repoPath.substring(1, repoPath.length()-1)
            echo(<span class="code-quote">"<span class="code-object">Thread</span> ${idx%threads}"</span>)
            checkouts[<span class="code-quote">"<span class="code-object">Thread</span> ${idx%threads}"</span>] &lt;&lt; [path: curatedRepoPath, url: <span class="code-quote">"${gitBaseUrl}/${curatedRepoUrl}"</span>]
            idx++
        }
    }
    stage(<span class="code-quote">'GitSlave Repos'</span>) {
        def doCheckouts = [:]
        <span class="code-keyword">for</span> (i = 0; i &lt; threads; i++){
            def j = i
            doCheckouts[<span class="code-quote">"<span class="code-object">Thread</span> ${j}"</span>] = {
                <span class="code-keyword">for</span> (co in checkouts[<span class="code-quote">"<span class="code-object">Thread</span> ${j}"</span>]) {
                    retry(3) {
                        checkout([$class: <span class="code-quote">'GitSCM'</span>,
                                branches: [[name: branch]],
                                doGenerateSubmoduleConfigurations: <span class="code-keyword">false</span>,
                                extensions: [[$class: <span class="code-quote">'RelativeTargetDirectory'</span>, relativeTargetDir: co.path], [$class: <span class="code-quote">'CleanCheckout'</span>], [$class: <span class="code-quote">'ScmName'</span>, name: co.path]],
                                submoduleCfg: [],
                                userRemoteConfigs: [[url: co.url]]])
                    }
                }
            }
        }
        parallel doCheckouts
    }
</pre>
</div></div></pre>
</details>
</details>
<details><summary><i>2 attachments</i></summary>

- [parallelcheckout.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/34959/parallelcheckout.png)
  > ![parallelcheckout.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/34959/parallelcheckout.png)
- [parallelcheckout2.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34958/parallelcheckout2.png)
  > ![parallelcheckout2.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34958/parallelcheckout2.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/roidelapluie.png" title="roidelapluie's avatar" /> roidelapluie:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=278184&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-278184">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>PS: I originally set this as major because of potential side effects of this bug. Maybe it affects all the SCM plugins, or affects other things than changelogs.</p>
</pre>
</details>

</ul>
</details>

<p>PS: I originally set this as major because of potential side effects of this bug. Maybe it affects all the SCM plugins, or affects other things than changelogs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=280909&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280909">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Could you explain more deeply what you were expecting instead of what you observed?  You said:</p>

<blockquote><p>One change is displayed in 8 different repositories, the 8 repositories that are checked out at the same time</p></blockquote>

<p>I interpret that to mean that you were expecting the change to only be found in zero or one of the eight repositories.  If so, then that is different than what I expect.  I expect that a single SHA1 is selected as the checkout for the duration of the job, unless the user specifically configures the definition in the checkout step to use something different.  If that isn't the case, then we risk that a user running a build may have a later checkout which fails to use the same SHA1 as the original checkout.</p>
</pre>
</details>

</ul>
</details>

<p>Could you explain more deeply what you were expecting instead of what you observed?  You said:</p>

<blockquote><p>One change is displayed in 8 different repositories, the 8 repositories that are checked out at the same time</p></blockquote>

<p>I interpret that to mean that you were expecting the change to only be found in zero or one of the eight repositories.  If so, then that is different than what I expect.  I expect that a single SHA1 is selected as the checkout for the duration of the job, unless the user specifically configures the definition in the checkout step to use something different.  If that isn't the case, then we risk that a user running a build may have a later checkout which fails to use the same SHA1 as the original checkout.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/roidelapluie.png" title="roidelapluie's avatar" /> roidelapluie:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=280910&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280910">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We check out 30 different git repositories.</p>

<p>We set the branch to 'master'.</p>

<p>The plugins pick the master branch of all the repositories.</p>

<p>But,</p>

<p>If there is a change in one repo, the plugin will display the change in all the repos that are checkout out at the same time (in other parallel() threads), even if the commit is not in that repo.</p>

</pre>
</details>

</ul>
</details>

<p>We check out 30 different git repositories.</p>

<p>We set the branch to 'master'.</p>

<p>The plugins pick the master branch of all the repositories.</p>

<p>But,</p>

<p>If there is a change in one repo, the plugin will display the change in all the repos that are checkout out at the same time (in other parallel() threads), even if the commit is not in that repo.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>pixman20:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=333688&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-333688">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This still appears to be a problem.</p>

<p>I noticed another problem today and finally realized several issues I've been seeing are related to Git not working with parallel steps in pipelines.</p>

<p>From what I can tell, the issue is that the changelog#.xml gets used for each checkout running at the same time.</p>

<p>For example, I have a job that does a parallel checkout of 3 different repositories.</p>

<p>On that run, each of them will write to changelog4.xml, which means the last one wins.</p>

<p>Also, in the build.xml for the run, each "org.jenkinsci.plugins.workflow.job.WorkflowRun_-SCMCheckout" node references changelog4.xml.</p>

<p>This is also probably why in the past sometimes links will go to the wrong repository.</p>

<p>At the time of the build what's in memory is fine, however after the memory is GC'ed and reloaded or Jenkins is restarted, the data from the changelog XMLs gets reloaded and (in my case) the changes for 1 repository are listed 3 times (1 for each of the different checkouts).</p>

<p> </p>

<p>I did a test and ran the same set of changes in another job that did a sequential checkout, and everything worked correctly, with each getting their own respective changelog#.xml.</p>

<p>FYI <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a></p>
</pre>
</details>

</ul>
</details>

<p>This still appears to be a problem.</p>

<p>I noticed another problem today and finally realized several issues I've been seeing are related to Git not working with parallel steps in pipelines.</p>

<p>From what I can tell, the issue is that the changelog#.xml gets used for each checkout running at the same time.</p>

<p>For example, I have a job that does a parallel checkout of 3 different repositories.</p>

<p>On that run, each of them will write to changelog4.xml, which means the last one wins.</p>

<p>Also, in the build.xml for the run, each "org.jenkinsci.plugins.workflow.job.WorkflowRun_-SCMCheckout" node references changelog4.xml.</p>

<p>This is also probably why in the past sometimes links will go to the wrong repository.</p>

<p>At the time of the build what's in memory is fine, however after the memory is GC'ed and reloaded or Jenkins is restarted, the data from the changelog XMLs gets reloaded and (in my case) the changes for 1 repository are listed 3 times (1 for each of the different checkouts).</p>

<p> </p>

<p>I did a test and ran the same set of changes in another job that did a sequential checkout, and everything worked correctly, with each getting their own respective changelog#.xml.</p>

<p>FYI <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=354288&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354288">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I just created and linked a new bug to this one (<a href="https://issues.jenkins.io/browse/JENKINS-54732" title="Revision information produced by pipline 'checkout' operation isn't parallel safe" class="issue-link" data-issue-key="JENKINS-54732">JENKINS-54732</a>), which I think may be related. My test case suggests that the SCM checkout function is not thread safe, sharing some global data or settings or something across invocations. This results in incorrect reference data describing the checkout to be propagated elsewhere - in my case back to the caller in the form of the data returned from the function call. That being the case this could explain why there are other problems / bugs like this one when running checkouts / clones in parallel.</p>
</pre>
</details>

</ul>
</details>

<p>I just created and linked a new bug to this one (<a href="https://issues.jenkins.io/browse/JENKINS-54732" title="Revision information produced by pipline 'checkout' operation isn't parallel safe" class="issue-link" data-issue-key="JENKINS-54732">JENKINS-54732</a>), which I think may be related. My test case suggests that the SCM checkout function is not thread safe, sharing some global data or settings or something across invocations. This results in incorrect reference data describing the checkout to be propagated elsewhere - in my case back to the caller in the form of the data returned from the function call. That being the case this could explain why there are other problems / bugs like this one when running checkouts / clones in parallel.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tsantos:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=355612&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-355612">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We are experiencing the same issue in our jenkins installation, while checking out many git repositories in parallel. 
 I tried to expose and fix the problem in this PR: <a href="https://github.com/jenkinsci/workflow-scm-step-plugin/pull/31" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-scm-step-plugin/pull/31</a></p>

<p>Hope you guys can review it (or fix it if there is a better way).</p>
</pre>
</details>

</ul>
</details>

<p>We are experiencing the same issue in our jenkins installation, while checking out many git repositories in parallel. <br/>
 I tried to expose and fix the problem in this PR: <a href="https://github.com/jenkinsci/workflow-scm-step-plugin/pull/31" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-scm-step-plugin/pull/31</a></p>

<p>Hope you guys can review it (or fix it if there is a better way).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-39968?focusedId=358470&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-358470">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/browse/JENKINS-54732" title="Revision information produced by pipline 'checkout' operation isn't parallel safe" class="issue-link" data-issue-key="JENKINS-54732">JENKINS-54732</a> seems distinct: about the return value of <tt>git</tt> / <tt>checkout</tt>; whereas IIUC this is about the build changelog, possibly making it a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-34313" title="SAXParseException from parallel SCM checkout" class="issue-link" data-issue-key="JENKINS-34313"><del>JENKINS-34313</del></a>.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/browse/JENKINS-54732" title="Revision information produced by pipline 'checkout' operation isn't parallel safe" class="issue-link" data-issue-key="JENKINS-54732">JENKINS-54732</a> seems distinct: about the return value of <tt>git</tt> / <tt>checkout</tt>; whereas IIUC this is about the build changelog, possibly making it a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-34313" title="SAXParseException from parallel SCM checkout" class="issue-link" data-issue-key="JENKINS-34313"><del>JENKINS-34313</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-34313%22">JENKINS-34313</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-55040%22">JENKINS-55040</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-54732%22">JENKINS-54732</a>]</i>
